### ISSUE TITLE
upgrade left half of cluster cordoned

### ISSUE THREAD

**Issue by snstanton** [bottlerocket-os/bottlerocket-update-operator#28]
**Image I'm using:**
328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-update-operator:v0.1.2

**Issue or Feature Request:**

After the recent release of Bottlerocket OS 0.3.1 my nodes started automatically upgrading as expected.  After a couple of hours, all of the nodes were upgraded but half of there were still in the `Ready,SchedulingDisabled` state.  I manually uncordoned the nodes and everything was find after that.

I have attached the pod logs for the controller and agents.

[pod.zip](https://github.com/bottlerocket-os/bottlerocket-update-operator/files/4315440/pod.zip)

**Comment by jahkeup** [comment]
Hi @snstanton, thanks for trying out the bottlerocket update operator!

I suspect that the controller is scheduled onto nodes and then immediately updated "off" of it (killed by kubelet) - while its processing other cluster nodes' intents.

Because the DeamonSets are scheduled on before uncordoning, the internal event loop of the controller may have written out its flag, saying it did so, without being able to complete the uncordon.

I'm away from my development machine at the moment, but once I'm back to it I'll take a peek at the code to see if my suspicions are true.

**Comment by jahkeup** [comment]
Okay, I found that there's a fair chance that the *agent* - not the controller - is what's failing to update its intent's status right before signaling for reboot ([the agent doesn't have retries built in right now](https://github.com/bottlerocket-os/bottlerocket-update-operator/blob/7fcaf306d983df8542bb31bcad68a9142cde82cf/pkg/agent/agent.go#L283-L285), but definitely should have em added!). I suspect that blocking progress on the successful update of the intent will prevent this particular issue from happening. I'm working on writing up a fix and repro case in testing to fix and protect against this behavior.

More thought needs to be given to the handling of the uncordoning of inprogress nodes - we can't assume that an cluster adminstrator did or didn't intentionally cordon off the node when the agent or controller observes this halfway state.

**Comment by jahkeup** [comment]
@snstanton I see what's going on now, I found that I had some nodes left cordoned when I put a large cluster through the same paces: some nodes would not be uncordoned and would also re-stabilize. This didn't happen every time and was closer to 20% of the time for me (3/16, 4/16, 3/16). What I found is not a race condition on the manipulation of annotations - instead, its is a race condition between the OS is killing processes for reboot and the agent observing its `updog` process.

The agent uses `updog` as a sort of API to the update process and this includes the step of actually signalling `init` (`systemd`) to reboot. `updog` however does not have a handler for the signal sent by `init` (`SIGTERM`) and instead exits with a non-zero status code. This status code is used by the agent as part of this informal API and the non-zero status code indicates failure or error. So, when this happens the agent sees the updater process exit updates its node's status with an "error" state. 

The agent posting an "error" for an observed "error" is indeed its expected behavior. If `updog` *did fail* then we do not want to proceed with anything without the controller telling the agent what to do next. However, *`updog` didn't fail and the node does reboot*. Before the reboot happens, the agent notes the error and updates the node's intent to reflect its "error" status.

We get these cordoned nodes because the controller will not bring a node out of the cordoned state without being confident that it's cordoned as part of the update process. Unfortunately, the combination of events (and posted intents) is wrongfully causing some nodes to remain cordoned for an administrator to inspect before workload returns.


I want to bounce a few ideas off of the team before outlining the next steps here - I'll post these ideas/thoughts and link related issues as they come. My guess is that there will be small, but related, changes in `updog` (in the OS image) and the agent.

---

This snippet from the gather logs shows `updog` being killed and reported as a failure - followed by the reboot:

https://gist.github.com/jahkeup/7eb651cb9dcb94de73b4804bd325e5b5#file-agent-pod-log-L9-L12
